File: ym/oa/valid_parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        tmp_stack = []
        dict = {"]":"[", "}":"{", ")":"("}
        for item in s:
            if item in dict.values():
                tmp_stack.append(item)
            elif item in dict.keys():
                if tmp_stack == [] or dict[item] != tmp_stack.pop():
                    return False
            else:
                return False
        return tmp_stack == []

    def test(self):
        input1 = '()'
        input2 = '(){}[]'
        input3 = '(]'
        input4 = '([)]'

        print input1 + ': ' + str(self.isValid(input1))
        print input2 + ': ' + str(self.isValid(input2))
        print input3 + ': ' + str(self.isValid(input3))
        print input4 + ': ' + str(self.isValid(input4))

File: ym/oa/ValidParentheses.java
import java.util.*;


class ValidParentheses {
    public static boolean isValid(String s) {
        Stack<Character> dicStack = new Stack<Character>();
        for (char ch : s.toCharArray()) {
            if (ch == '(')
                dicStack.push(')');
            else if (ch == '{')
                dicStack.push('}');
            else if (ch == '[')
                dicStack.push(']');
            else if (dicStack.isEmpty() || dicStack.pop() != ch)
                return false;
        }
        return dicStack.isEmpty();
    }

    public static void main(String[] args){
        String input1 = "()";
        String input2 = "(){}[]";
        String input3 = "(]";
        String input4 = "([)]";

        String[] input = {
                "()", "(){}[]", "(]", "([)]"
        };

        for (int i = 0; i < input.length; i++){
            System.out.println(input[i] + ": " + isValid(input[i]));
        }
    }
}